### ISSUE TITLE
Two questions regarding model implementation: 

### ISSUE THREAD

**Issue by vinsis** [openai/CLIP#48]
First of all thank you for providing the code and pretrained model. The results look stunning.

I was looking into the source code and there are two parts I am struggling to understand:

1. What is the purpose of `logit_scale` in the model [here](https://github.com/openai/CLIP/blob/main/clip/model.py#L362)? It seems to simply scale the dot products between two representations.
2. In the implementation of `VisualTransformer` what is the role of `class_embedding` defined [here](https://github.com/openai/CLIP/blob/main/clip/model.py#L209).

Thanks!

**Comment by PaulLerner** [comment]
>     2. In the implementation of `VisualTransformer` what is the role of `class_embedding` defined [here](https://github.com/openai/CLIP/blob/main/clip/model.py#L209).

This is the embedding used for the final representation, see [here](https://github.com/openai/CLIP/blob/main/clip/model.py#L230) they only keep the first "token" of each sequence.
See http://arxiv.org/abs/2010.11929 for more details about Vision Transformer :)

**Comment by vinsis** [comment]
Got it, thanks. 

The answer to the first question still remains elusive.

**Comment by jongwook** [comment]
During contrastive learning, we interpret the dot products as logits that are fed to the cross-entropy loss for each text and image example. Because we normalize the feature vectors, the dot products are capped between [-1, 1], which might not have the enough dynamic range and limit the categorical probability distributions that the logits can express. For this reason, we scale the logits to have larger differences. To avoid numerical instability, we clipped the scale value at 100, and for all models that we trained, it reached 100. [This comment](https://github.com/openai/CLIP/issues/46#issuecomment-782558799) explains how to clip the value.

**Comment by vinsis** [comment]
Ok thanks. That explains it.

**Comment by Horizon2333** [comment]
> During contrastive learning, we interpret the dot products as logits that are fed to the cross-entropy loss for each text and image example. Because we normalize the feature vectors, the dot products are capped between [-1, 1], which might not have an enough dynamic range and limit the categorical probability distributions that the logits can express. For this reason, we scale the logits to have larger differences. To avoid numerical instability, we clipped the scale value at 100, and for all models that we trained, it reached 100. [This comment](https://github.com/openai/CLIP/issues/46#issuecomment-782558799) explains how to clip the value.

So why don't you directly initialize temperature parameter to 100 ?  What's the purpose of let it grow from 1/0.07 to 100? Does it  bring enough performance improvement?

**Comment by thesby** [comment]
> > During contrastive learning, we interpret the dot products as logits that are fed to the cross-entropy loss for each text and image example. Because we normalize the feature vectors, the dot products are capped between [-1, 1], which might not have an enough dynamic range and limit the categorical probability distributions that the logits can express. For this reason, we scale the logits to have larger differences. To avoid numerical instability, we clipped the scale value at 100, and for all models that we trained, it reached 100. [This comment](https://github.com/openai/CLIP/issues/46#issuecomment-782558799) explains how to clip the value.
> 
> So why don't you directly initialize temperature parameter to 100 ? What's the purpose of let it grow from 1/0.07 to 100? Does it bring enough performance improvement?

Yes, I think initialize to 100 will converge faster. But why not set to 100?

**Comment by yakorch** [comment]
@Horizon2333 @thesby
I am late to the party, but perhaps this is still useful.
In [Temperature-Free Loss Function for Contrastive Learning](https://arxiv.org/pdf/2501.17683v1), the authors analyze the properties of the InfoNCE loss, showing that for high temperature values (in the paper notation that I cite), the total loss is non-zero even for correct model predictions (C -> 1).

However, my intuition is that at the start of the training, the model is poor, resulting in very few "C->1" predictions, so the optimization process does not experience this drawback of the chosen loss function. And this comes with a neat bonus of a strong gradient magnitude for non-optimal C, speeding up training.

After some training, hopefully the model is better, and "C->1" predictions are common, and one can no longer ignore the non-zero InfoNCE loss. Yet it is no longer a problem thanks to the decreasing temperature as a learnable parameter.
